### ISSUE TITLE
Q: Guidelines for mutating uplot instance vs destroying & recreating a new one

### ISSUE THREAD

**Issue by FilipLeonard** [leeoniya/uPlot#836]
Hi Leon, thanks again for this nice piece of software.

What are some soft guidelines that you generally consider when implementing _plot updates_? 

Simply thinking in black and white, I can see two high-level strategies: _either_  work directly with the uplot instance and mutate the various options (e.g. `uplot.setSeries(idx, {...uplot.series[idx], [updatedSeriesAttribute]: newValue})` ), and only recreate(destroy+create) the instance when absolutely needed, _or_  maintain some sort of instance/model for the options (e.g. something like a builder pattern) and whenever those options change in any way, immediately recreate the plot. 
Of course, actual coding comes in a lot of greys and I can easily see some cases where one of the patterns would make less sense (e.g. when hiding one of the series, why recreate the entire chart with updated options when you can just toggle it on the plot itself). Since you're familiar with the inner workings of the library and its performance, I was wondering if you follow some patterns or soft rules in that regard, for instance, for a set of actions, you always update the chart and for another set, you always destroy and recreate it. Much appreciated!

**Comment by leeoniya** [comment]
hey!

`uplot.setSeries()` does not support updating arbitrary properties, it's mostly there to programmatically toggle show/focus (as what the legend does).

>  (e.g. something like a builder pattern) and whenever those options change in any way, immediately recreate the plot.

this is what we do in Grafana. we use a builder pattern [1] with options diffing for full re-init. i think this is the right choice because it's not worth hyper-optimizing for opts that change infrequently, or only during initial panel edit/setup. even the show/hide in Grafana's legends, as well as panel config sliders [3] do a full re-init from `oninput` events.

only stuff that is likely to happen at runtime, and multiple times per second, should try to avoid re-init [2], like mousemove hover/focus, rapid data updates, plot resizing.

you could use various callbacks in uPlot options and rely on closures to return mutable/dynamic stuff e.g. `series.stroke()`, etc, and/or carefully track series and use `.addSeries() / .delSeries()`, but it just ends up being a bunch of extra fragile code to maintain for very little real benefit.

[1] https://github.com/grafana/grafana/tree/main/packages/grafana-ui/src/components/uPlot/config
[2] https://github.com/grafana/grafana/blob/6c7b17b59fa77d8413b90f6395b057e1f591009c/packages/grafana-ui/src/components/uPlot/Plot.tsx#L91-L109
[3] 
![image](https://github.com/leeoniya/uPlot/assets/43234/58b280dd-2617-4366-89cb-f9d35330972b)

**Comment by FilipLeonard** [comment]
Hey!

Thanks a lot!

> even the show/hide in Grafana's legends, as well as panel config sliders [3] do a full re-init from `oninput` events

> only stuff that is likely to happen at runtime, and multiple times per second, should try to avoid re-init [2], like mousemove hover/focus, rapid data updates, plot resizing.

These are some spot-on answers.

I'd like to ask just one more thing. How strict are you with what you put inside `data`? Do you religiously put in there _only_ the actual metrics data, and nothing else? If yes, is there a perf reason for this?

I've found that in given situations where other secondary, assistive lines (e.g. a threshold, a reference metric) need to be drawn along with the main lines, putting these pseudo series in `data` makes for a quick win, but feels dirty... I noticed that in Grafana you draw the thresholds separately, directly with the `ctx` reference, but I was curious if you also have a guideline for this. Thanks again!

**Comment by leeoniya** [comment]
> I'd like to ask just one more thing. How strict are you with what you put inside `data`? Do you religiously put in there only the actual metrics data, and nothing else? If yes, is there a perf reason for this?

yeah, this is a bit of a wart that i'd like to fix for v2.0.

sometimes i'll stick it onto the end of `data` and omit from series defs. sometimes i'll have a series def if i want it in legend and toggle-able. it doesnt need to be aligned if you implement your own `series.paths` renderer or use hooks to access that data and render it as needed. in grafana sometimes we'll have plugin that keeps the un-alignable data externally and is rendered from the scope of the plugin where the uplot hooks have access to it. it works, is fast, but is ungreat.

for v2, the plan is to move exclusively to an API that has an extra array layer, so that you can have multiple datasets that don't need to be cross-aligned on x. that way you can stick annotations, thresholds, regions, etc. into the data without it feeling too awkward.

current:

```js
let data = [
  [0,1,2], // shared x
  [5,5,5], // y1
  [9,9,9], // y2
]
```

v2:

```js
let data = [
  [
    [0,1,2], // x1
    [5,5,5], // y1
  ],
  [
    [5,6,7], // x2
    [9,9,9], // y2
  ]
];
```

v2 will detect if `x1 === x2` (by reference), and will apply the same optimizations that exist today in the single/aligned dataset. still working out exact semantics/rules for this ;)

sorry for late reply.

**Comment by FilipLeonard** [comment]
Hey @leeoniya, thanks, appreciate the response!
